yt/frontends/stream/data_structures.py

```diff
@@ -1092,11 +1138,81 @@ def __init__(self, stream_handler, storage_filename=None,
         if ('io', 'density') in fields and ('io', 'smoothing_length') in fields:
             self._sph_ptype = 'io'
 
-def load_particles(data, length_unit = None, bbox=None,
-                   sim_time=0.0, mass_unit = None, time_unit = None,
+    def add_sph_fields(self, n_neighbors=64, sph_ptype="io"):
+        """Add SPH fields for the specified particle type.
+
+        For a particle type with "particle_position" and "particle_mass" already
+        defined, this method add the "smoothing_length" and "density" fields.
+        "smoothing_length" is computed as the distance to the nth nearest
+        neighbor. "density" is simply assuming a sphere with "smoothing_length"
+        as its radius.
+
+        Parameters
+        ----------
+        n_neighbors : int
+            The number of neighbors to use when computing smoothing length.
+        sph_ptype : str
+            The SPH particle type. Each dataset could have one SPH particle type
+            only.
+
+        """
+        ad = self.all_data()
+
+        l_unit = "code_length"
+        m_unit = "code_mass"
+        d_unit = "code_mass / code_length**3"
+
+        # Read basic particle fields
+        pos = ad[sph_ptype, "particle_position"].to(l_unit).v
+        mass = ad[sph_ptype, "particle_mass"].to(m_unit).v
+
+        def exists(fname):
+            if (sph_ptype, fname) in self.derived_field_list:
+                mylog.info("Field ('%s','%s') already exists. Skipping",
+                            sph_ptype, fname)
+                return True
+            else:
+                mylog.info("Generating field ('%s','%s')",
+                            sph_ptype, fname)
+                return False
+
+        data = {}
+
+        # Compute smoothing length field
+        fname = "smoothing_length"
+        if not exists(fname):
+            left_edge = self.domain_left_edge.to(l_unit).v
+            right_edge = self.domain_right_edge.to(l_unit).v
+            kdtree = PyKDTree(
+                pos.astype("float64"),
+                left_edge=left_edge,
+                right_edge=right_edge,
+                periodic=self.periodicity,
+                leafsize=2*int(n_neighbors),
+            )
+            pos_kd = pos[kdtree.idx]
+            hsml = generate_smoothing_length(pos_kd, kdtree, n_neighbors)
+            hsml = hsml[np.argsort(kdtree.idx)]
+            data[(sph_ptype, fname)] = (hsml, l_unit)
+
+        # Compute density field
```

Comment:
I need some help with the computation of density given mass and smoothing length. There are issues at two levels:

1. The most accurate SPH density estimation requires a non-local operation, which is not readily available in yt yet. Do we want to go in that direction and add the tool?

2. It's probably okay we don't do the most accurate computation here. Because the user could provide the density field themselves if this is important. We just need to make it clear in the docs what we are doing here. `dens = mass / vol` is correct in dimension, but might be wrong in normalization, which is dependant on the smoothing kernel. My proposal in this direction is to add a `normalization_factor` argument to this method, which defaults to the case of the cubic spline.

I couldn't make a decision on myself. So I'm looking forward to having more discussions with people.